server/src/internalClusterTest/java/org/elasticsearch/action/admin/cluster/node/tasks/TasksIT.java

```diff
@@ -599,22 +600,27 @@ private <T> void waitForCompletionTestCase(boolean storeResult, Function<TaskId,
             TEST_TASK_ACTION.name() + "[n]",
             () -> client().execute(TEST_TASK_ACTION, request)
         );
-        ActionFuture<T> waitResponseFuture;
-        try {
-            var tasks = clusterAdmin().prepareListTasks().setActions(TEST_TASK_ACTION.name()).get().getTasks();
-            assertThat(tasks, hasSize(1));
-            var taskId = tasks.get(0).taskId();
-            clusterAdmin().prepareGetTask(taskId).get();
-
-            // Spin up a request to wait for the test task to finish
-            waitResponseFuture = wait.apply(taskId);
-        } finally {
-            // Unblock the request so the wait for completion request can finish
-            client().execute(UNBLOCK_TASK_ACTION, new TestTaskPlugin.UnblockTestTasksRequest()).get();
-        }
-
-        // Now that the task is unblocked the list response will come back
-        T waitResponse = waitResponseFuture.get();
+        var tasks = clusterAdmin().prepareListTasks().setActions(TEST_TASK_ACTION.name()).get().getTasks();
+        assertThat(tasks, hasSize(1));
+        TaskId taskId = tasks.get(0).taskId();
+        clusterAdmin().prepareGetTask(taskId).get();
+
+        var taskManager = (MockTaskManager) internalCluster().getInstance(
+            TransportService.class,
+            clusterService().state().getNodes().resolveNode(taskId.getNodeId()).getName()
+        ).getTaskManager();
+        taskManager.addListener(new MockTaskManagerListener() {
+            @Override
+            public void onRemovedTaskListenerRegistered(RemovedTaskListener removedTaskListener) {
+                // Unblock the request only after it started waiting for task completion
+                if (removedTaskListener.toString().startsWith("Completing running task Task{id=" + taskId.getId())) {
```

Comment:
This seems a bit strange, I think it works without it too, since there should be no other wait for completions going on.

Reply:
@henningandersen There seems to be a bug in `TestTaskPlugin#TransportTestTaskAction`. It checks whether a task is blocked by running `waitUntil` for 10 seconds, but doesn't check whether  `waitUntil` finished successfully.